lib/std/heap/general_purpose_allocator.zig

```diff
@@ -0,0 +1,920 @@
+            const self = @fieldParentPtr(Self, "allocator", allocator);
+
+            const held = self.mutex.acquire();
+            defer held.release();
+
+            const prev_req_bytes = self.total_requested_bytes;
+            if (config.enable_memory_limit) {
+                const new_req_bytes = prev_req_bytes + len;
+                if (new_req_bytes > self.requested_memory_limit) {
+                    return error.OutOfMemory;
+                }
+                self.total_requested_bytes = new_req_bytes;
+            }
+            errdefer if (config.enable_memory_limit) {
+                self.total_requested_bytes = prev_req_bytes;
+            };
+
+            const new_aligned_size = math.max(len, ptr_align);
+            if (new_aligned_size > largest_bucket_object_size) {
+                try self.large_allocations.ensureCapacity(
+                    self.backing_allocator,
+                    self.large_allocations.entries.items.len + 1,
+                );
+
+                const slice = try self.backing_allocator.allocFn(self.backing_allocator, len, ptr_align, len_align);
+
+                const gop = self.large_allocations.getOrPutAssumeCapacity(@ptrToInt(slice.ptr));
+                assert(!gop.found_existing); // This would mean the kernel double-mapped pages.
+                gop.entry.value.bytes = slice;
+                collectStackTrace(@returnAddress(), &gop.entry.value.stack_addresses);
+
+                return slice;
+            } else {
+                const new_size_class = up_to_nearest_power_of_2(usize, new_aligned_size);
+                const ptr = try self.allocSlot(new_size_class, @returnAddress());
+                return ptr[0..len];
+            }
+        }
+
+        fn createBucket(self: *Self, size_class: usize, bucket_index: usize) Error!*BucketHeader {
+            const page = try self.backing_allocator.allocAdvanced(u8, page_size, page_size, .exact);
+            errdefer self.backing_allocator.free(page);
+
+            const bucket_size = bucketSize(size_class);
+            const bucket_bytes = try self.backing_allocator.allocAdvanced(u8, @alignOf(BucketHeader), bucket_size, .exact);
+            const ptr = @ptrCast(*BucketHeader, bucket_bytes.ptr);
+            ptr.* = BucketHeader{
+                .prev = ptr,
+                .next = ptr,
+                .page = page.ptr,
+                .alloc_cursor = 0,
+                .used_count = 0,
+            };
+            self.buckets[bucket_index] = ptr;
+            // Set the used bits to all zeroes
+            @memset(@as(*[1]u8, ptr.usedBits(0)), 0, usedBitsCount(size_class));
+            return ptr;
+        }
+    };
+}
+
+const TraceKind = enum {
+    alloc,
+    free,
+};
+
+fn up_to_nearest_power_of_2(comptime T: type, n: T) T {
+    var power: T = 1;
+    while (power < n)
+        power *= 2;
+    return power;
+}
+
+fn hash_addr(addr: usize) u32 {
```

Comment:
`AutoHashMap` should be better, especially now that it has a fastpath for easy to hash keys.